### ISSUE TITLE
Negative loss value of nn.KLDivLoss

### ISSUE THREAD

**Issue by songyuc** [pytorch/pytorch#60928]
## 🐛 Bug

I got a negative loss value when using `nn.KLDivLoss` 

## To Reproduce

1. use the  code as,
```python
import numpy as np
import torch
import torch.nn as nn
loss = nn.KLDivLoss()
output = torch.from_numpy(np.array([[0.1132, 0.5477, 0.3390]])).float()
target = torch.from_numpy(np.array([[0.1541, 0.0511, 0.0947]])).float()
loss(output.log(),target)
```
2. get the loss as "tensor(-0.0648)", which is negative.

Here is the test code,
```python
import numpy as np
import torch
import torch.nn as nn
loss = nn.KLDivLoss()
output = torch.from_numpy(np.array([[0.1132, 0.5477, 0.3390]])).float()
target = torch.from_numpy(np.array([[0.1541, 0.0511, 0.0947]])).float()
loss(output.log(),target)
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
Get the information as,
```bash
OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.10

Python version: 3.7.3 (default, Mar 27 2019, 22:11:17)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-77-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: N/A
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 460.80
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.1
[pip3] torchaudio==0.7.0a0+a853dff
[pip3] torchvision==0.8.0a0
[conda] blas                      1.0                         mkl    defaults
[conda] mkl                       2019.4                      243    defaults
[conda] mkl-service               2.0.2            py37h7b6447c_0    defaults
[conda] mkl_fft                   1.0.12           py37ha843d7b_0    defaults
[conda] mkl_random                1.0.2            py37hd81dba3_0    defaults
[conda] numpy                     1.16.4           py37h7e9f1db_0    defaults
[conda] numpy-base                1.16.4           py37hde5b4d6_0    defaults
[conda] numpydoc                  0.9.1                      py_0    defaults
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by jbschlosser** [comment]
Hey @songyuc, note that, as specified in the [docs](https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html), `KLDivLoss` expects `output` to be log-probabilities and `target` to be probabilities. Tweaking your example above to ensure this gives a positive loss value:

```python
import torch
import torch.nn as nn

loss = nn.KLDivLoss()
output = torch.tensor([[0.1132, 0.5477, 0.3390]]).log_softmax(dim=1)
target = torch.tensor([[0.1541, 0.0511, 0.0947]]).softmax(dim=1)
print(loss(output, target))
```
```
tensor(0.0080)
```

**Comment by songyuc** [comment]
Hi, @jbschlosser , is the `log_softmax` essential for `output` tensor?

**Comment by jbschlosser** [comment]
Hey @songyuc, the key thing is that `output` should contain log-probabilities:
* If `output` contains raw outputs from e.g. a `Linear` layer (often referred to as "logits"), the logits can be converted to log-probabilities using `log_softmax()`
* If `output` comes from a `Softmax` layer (interpretable as probabilities), the probabilities can be converted to log-probabilities using `log()`

Hope that helps!
